docs/proposals/2018-05-declarative-delete-field.md

```diff
@@ -0,0 +1,86 @@
+# Declarative Delete Field
+
+Concept committed to [during 2018 April 12F2F](https://docs.google.com/document/d/1O7_fws7hwZ6qV3okAjbV5qdFaEq4bt1LYJ9bTu-IwvU)
+
+This document 2018-05.
+
+## Abstract
+
+Add a field to instance and binding to indicate that they have been asked to
+send a DELETE to the backend.
+
+## Motivation
+
+When the Originating-Identity header is used, the broker and refuse a delete
+call when the user info in the Originating-Identity header does not have
+permissions to delete the instance or binding. As this happens after kubernetes
+has accepted the DELETE as an true indication to do a delete, it has set the
+DeletionTimestamp, which is an irreversible step on the way to carrying out the
+necessary actions to delete a kubernetes resource.
+
+## Proposed Design
+
+ - Add a field to indicate deletion
+ - Add a condition to indicate permanent end state of deletion success.
+
+Once the field is activated the controller deprovisions the broker
+resource. When successfully deleted the controller then deletes the
```

Comment:
What if deprovision via OSB fails? There are various cases, e.g.:
- initial request is rejected with `400 BadRequest`, `403 Forbidden` or some other 4xx client-side error that should have no side effects
- initial request ends with `5xx` server error
- initial request processing times out
- initial request is accepted, triggering async deprovisioning and `last_operation` returns failure

Reply:
Isn't this the normal cases that are already handled? We're doing a deprovision outside of calling delete. Everything else about the broker operation and responses is the same.

Reply:
I don't understand. What's the point then? What do you achieve by doing exactly the same as with normal delete?

As far as I understand, the only reason for adding new flag is because you can't "undelete" Kubernetes resource: once "deletionTimestamp" is added, you can't remove it and can't change the spec.
Do you unset the flag or allow updating the spec in some cases? If not then I don't see the point of having it.

Reply:
To clarify, in case you don't know. Currently when user marks `ServiceInstance` for deletion, it doesn't get deleted by GC immediately, because we have a custom "Service Catalog finalizer" added to it. Once deprovisioning succeeds, svc-cat controller removes the finalizer and GC proceeds with deletion. If not, we retry until exceed limits, and then get stuck in the "failed to deprovision" state (and the only way to delete this ServiceInstance resource is to manually delete the finalizer).

It looks like you try to emulate exactly the same mechanics in your design. Why???

Reply:
Because, once the deletion timestamp is added, it cannot be removed. The bool can.

Reply:
What happens if you remove it in the middle of deprovisioning?

Reply:
Depends on when you mean in the middle. If it's after the deprovision call, that is synchronous and is a yes or no. If we catch it before the deprovision, it wouldn't do it. And if we catch it after, it's already done. We should prevent it from being unset, but the key thing is that the status changes to some form of "gone, permanently". I never proposed and do not intend to implement or support flipping back and causing a (re)provision.

Reply:
Right, so then you'll need to add some extra flag like `permanentDeletion bool` to the status.